pkg/gardenadm/translator/static_pod.go

```diff
@@ -0,0 +1,153 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+// Translate translates the given object into a list of files containing static pod manifests as well as ConfigMaps and
+// Secrets that can be injected into an OperatingSystemConfig.
+func Translate(ctx context.Context, c client.Client, o client.Object) ([]extensionsv1alpha1.File, error) {
+	switch obj := o.(type) {
+	case *appsv1.Deployment:
+		return translatePodTemplate(ctx, c, obj.ObjectMeta, obj.Spec.Template)
+	// TODO(rfranzke): Consider adding support for StatefulSet in the future.
+	default:
+		return nil, fmt.Errorf("unsupported object type %T", o)
+	}
+}
+
+func translatePodTemplate(ctx context.Context, c client.Client, objectMeta metav1.ObjectMeta, podTemplate corev1.PodTemplateSpec) ([]extensionsv1alpha1.File, error) {
+	pod := &corev1.Pod{ObjectMeta: podTemplate.ObjectMeta, Spec: podTemplate.Spec}
+	pod.Name = objectMeta.Name
+	pod.Namespace = objectMeta.Namespace
+
+	translateConfig(&pod.Spec)
+	translateContainers(pod.Spec.Containers)
+
+	filesFromVolumes, err := translateVolumes(ctx, c, pod)
+	if err != nil {
+		return nil, fmt.Errorf("failed translating volumes for static pod %s: %w", client.ObjectKeyFromObject(pod), err)
+	}
+
+	staticPodYAML, err := kubernetesutils.Serialize(pod, c.Scheme())
+	if err != nil {
+		return nil, fmt.Errorf("failed serializing static pod manifest for %s to YAML: %w", client.ObjectKeyFromObject(pod), err)
+	}
+
+	return append([]extensionsv1alpha1.File{{
+		Path:        filepath.Join("/", "etc", "kubernetes", "manifests", pod.Name+".yaml"),
+		Permissions: ptr.To[uint32](0644),
+		Content:     extensionsv1alpha1.FileContent{Inline: &extensionsv1alpha1.FileContentInline{Data: staticPodYAML}},
+	}}, filesFromVolumes...), nil
+}
+
+func translateConfig(spec *corev1.PodSpec) {
+	spec.HostNetwork = true
+	spec.PriorityClassName = "system-node-critical"
+}
+
+func translateContainers(containers []corev1.Container) {
+	for i, container := range containers {
+		if container.LivenessProbe != nil && container.LivenessProbe.HTTPGet != nil {
+			containers[i].LivenessProbe.HTTPGet.Host = "127.0.0.1"
+		}
+		if container.ReadinessProbe != nil && container.ReadinessProbe.HTTPGet != nil {
+			containers[i].ReadinessProbe.HTTPGet.Host = "127.0.0.1"
+		}
+		if container.StartupProbe != nil && container.StartupProbe.HTTPGet != nil {
+			containers[i].StartupProbe.HTTPGet.Host = "127.0.0.1"
+		}
+	}
+}
+
+func translateVolumes(ctx context.Context, c client.Client, pod *corev1.Pod) ([]extensionsv1alpha1.File, error) {
+	var (
+		files               []extensionsv1alpha1.File
+		addFileWithHostPath = func(hostPath, fileName, content string, permissions *int32, desiredItems []corev1.KeyToPath) {
```

Comment:
Why not use an ordinary function instead of specifying it here in line? There is no variance. Hence, a static function call could be sufficient.

Reply:
This adds directly to the `files` list if certain conditions are met. Making this a "regular" function outside would result in more complexity IMO. While I'm generally a fan of "regular" functions when applicable, I think here it's a benefit to define it this way.